presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,229 @@
+        {
+            List<PlanNode> sources = node.getSources();
+            int numberOfSources = sources.size();
+
+            ImmutableList.Builder<PlanNode> newProjectNodes = ImmutableList.builder();
+            ImmutableListMultimap.Builder<Symbol, Symbol> symbolMappings = ImmutableListMultimap.builder();
+            List<Symbol> unionOutputSymbols = null;
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregationsBuilder = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Expression> outputProjections = ImmutableMap.builder();
+
+            for (int i = 0; i < numberOfSources; i++) {
+                ProjectNode newProjectNode = createProjectNode(node, sources, i);
+                newProjectNodes.add(newProjectNode);
+
+                List<Symbol> projectOutputSymbols = newProjectNode.getOutputSymbols();
+                if (unionOutputSymbols == null) {
+                    // Use the first Relation to derive output symbol names
+                    unionOutputSymbols = getUnionOutputSymbolsFrom(projectOutputSymbols);
+                }
+
+                checkArgument(projectOutputSymbols.size() == unionOutputSymbols.size(),
+                        "Intersect output should have %s symbols but has %s symbols",
+                        unionOutputSymbols.size(),
+                        projectOutputSymbols.size());
+
+                int unionFieldId = 0;
+                for (Symbol symbol : projectOutputSymbols) {
+                    symbolMappings.put(unionOutputSymbols.get(unionFieldId), symbol);
+                    unionFieldId++;
+                }
+            }
+
+            UnionNode unionNode = new UnionNode(node.getId(), newProjectNodes.build(), symbolMappings.build(), ImmutableList.copyOf(symbolMappings.build().keySet()));
+            AggregationNode aggregationNode = createAggregationNode(node, unionOutputSymbols, unionNode, aggregationsBuilder, outputProjections);
+            FilterNode filterNode = createFilterNode(aggregationNode, aggregationsBuilder.build());
+
+            return new ProjectNode(idAllocator.getNextId(), filterNode, outputProjections.build());
+        }
+
+        private List<Symbol> getUnionOutputSymbolsFrom(List<Symbol> sourceSymbols)
+        {
+            ImmutableList.Builder<Symbol> outputSymbolBuilder = ImmutableList.builder();
+            for (Symbol symbol : sourceSymbols) {
+                outputSymbolBuilder.add(allocateNewSymbolFrom(symbol));
+            }
+            return outputSymbolBuilder.build();
+        }
+
+        private Symbol allocateNewSymbolFrom(Symbol symbol)
+        {
+            return allocateNewSymbolFrom(symbol.getName(), symbolAllocator.getTypes().get(symbol));
+        }
+
+        private AggregationNode createAggregationNode(IntersectNode node, List<Symbol> outputSymbols, UnionNode sourceNode, ImmutableMap.Builder<Symbol, FunctionCall> aggregationsBuilder, ImmutableMap.Builder<Symbol, Expression> outputProjections)
```

Comment:
can you call this something like createAggregationNodeForIntersect()? I was confused when first reading through because I didn't see the group by things added.